xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -418,46 +481,77 @@ public Map<ServerInfo, LoadReportClient> getServerLrsClientMap() {
     return ImmutableMap.copyOf(serverLrsClientMap);
   }
 
-  @Nullable
-  private ServerInfo getServerInfo(String resource) {
+  private String getAuthority(String resource) {
+    String authority;
     if (resource.startsWith(XDSTP_SCHEME)) {
       URI uri = URI.create(resource);
-      String authority = uri.getAuthority();
+      authority = uri.getAuthority();
       if (authority == null) {
         authority = "";
       }
+    } else {
+      authority = null;
+    }
+
+    return authority;
+  }
+
+  @Nullable
+  private ImmutableList<ServerInfo> getServerInfos(String authority) {
+    if (authority != null) {
       AuthorityInfo authorityInfo = bootstrapInfo.authorities().get(authority);
       if (authorityInfo == null || authorityInfo.xdsServers().isEmpty()) {
         return null;
       }
-      return authorityInfo.xdsServers().get(0);
+      return authorityInfo.xdsServers();
     } else {
-      return bootstrapInfo.servers().get(0); // use first server
+      return bootstrapInfo.servers();
+    }
+  }
+
+  private List<String> getAuthoritiesForServerInfo(ServerInfo serverInfo) {
+    List<String> authorities = new ArrayList<>();
+    for (Map.Entry<String, AuthorityInfo> entry : bootstrapInfo.authorities().entrySet()) {
+      if (entry.getValue().xdsServers().contains(serverInfo)) {
+        authorities.add(entry.getKey());
+      }
+    }
+
+    if (bootstrapInfo.servers().contains(serverInfo)) {
+      authorities.add(null);
     }
+
+    return authorities;
   }
 
   @SuppressWarnings("unchecked")
   private <T extends ResourceUpdate> void handleResourceUpdate(
       XdsResourceType.Args args, List<Any> resources, XdsResourceType<T> xdsResourceType,
       ProcessingTracker processingTracker) {
+    ControlPlaneClient controlPlaneClient = serverCpClientMap.get(args.serverInfo);
+
     ValidatedResourceUpdate<T> result = xdsResourceType.parse(args, resources);
     logger.log(XdsLogger.XdsLogLevel.INFO,
         "Received {0} Response version {1} nonce {2}. Parsed resources: {3}",
-         xdsResourceType.typeName(), args.versionInfo, args.nonce, result.unpackedResources);
+        xdsResourceType.typeName(), args.versionInfo, args.nonce, result.unpackedResources);
     Map<String, ParsedResource<T>> parsedResources = result.parsedResources;
     Set<String> invalidResources = result.invalidResources;
     List<String> errors = result.errors;
     String errorDetail = null;
     if (errors.isEmpty()) {
       checkArgument(invalidResources.isEmpty(), "found invalid resources but missing errors");
-      serverCpClientMap.get(args.serverInfo).ackResponse(xdsResourceType, args.versionInfo,
+      controlPlaneClient.ackResponse(xdsResourceType, args.versionInfo,
           args.nonce);
+      if (controlPlaneClient.isStartingUp()) {
+        controlPlaneClient.startingUpCompleted();
+        shutdownLowerPriorityCpcs(controlPlaneClient);
```

Comment:
This looks wrong being only for if `errors.isEmpty()`. Why is it dependent on whether we are ACKing or NACKing?

Reply:
done